File: DemoApp/DemoApp/MasterTableViewController.m
//
//  AVOMasterTableViewController.m
//  DemoApp
//
//  Created by Glispa GmbH on 03/12/15.
//  Copyright © 2015 Glispa GmbH All rights reserved.
//

#import "MasterTableViewController.h"
#import "MasterTableItem.h"
#import "AppControllerManagement.h"

@interface MasterTableViewController ()

@end

@implementation MasterTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

#pragma mark - Public

- (void)update{
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.adItems.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];

    MasterTableItem *item = self.adItems[indexPath.row];
    cell.textLabel.font = [UIFont fontWithName:@"Helvetica" size:14];
    cell.textLabel.textColor = [UIColor darkGrayColor];
    cell.textLabel.text = item.title;

    return cell;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    MasterTableItem *item = self.adItems[indexPath.row];
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *viewController;
    if (item.controller && item.xibName) {
        Class class = NSClassFromString(item.controller);
        viewController = [[class alloc] initWithNibName:item.xibName bundle:[NSBundle mainBundle]];
        [self.navigationController showViewController:viewController sender:nil];
    } else if (item.controller) {
        viewController = [storyboard instantiateViewControllerWithIdentifier:item.controller];
        [self.navigationController pushViewController:viewController animated:YES];
    } else if (item.childItems.count > 0) {
        MasterTableViewController *tableController = (MasterTableViewController*)[storyboard instantiateViewControllerWithIdentifier:@"MasterTableViewController"];
        tableController.adItems = item.childItems;
        tableController.title = item.title;
        [self.navigationController pushViewController:tableController animated:YES];
        viewController = tableController;
    }
    
    if ([viewController conformsToProtocol:@protocol(AppControllerManagement)] &&
        [viewController respondsToSelector:@selector(applyUserData:)]) {
        [(UIViewController<AppControllerManagement> *)viewController applyUserData:item.userData];
    }
}

@end

File: DemoApp/DemoAppSwift/MasterTableViewController.swift
//
//  MasterTableViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 19/05/16.
//  Copyright © 2016 Glispa GmbH. All rights reserved.
//

import UIKit

class MasterTableViewController: UITableViewController {

    open var adItems: [MasterTableItem]?

    open func update() {
        self.tableView.reloadData()
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return adItems?.count ?? 0
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)

        cell.textLabel?.font = UIFont(name: "Helvetica", size: 16)
        cell.textLabel?.textColor = UIColor.darkGray
        cell.textLabel?.text = adItems?[(indexPath as NSIndexPath).row].title

        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let item = adItems?[(indexPath as NSIndexPath).row] else { return }
        let storyboard: UIStoryboard = UIStoryboard(name: "Main", bundle: nil)
        var viewController: UIViewController? = nil
        var showBlock: (() -> Void)!
        if let xibName = item.xibName, let controllerName = item.controller {
            let viewControllerClass = NSClassFromString(controllerName) as! UIViewController.Type
            viewController = viewControllerClass.init(nibName: xibName, bundle: Bundle.main)
            showBlock = {self.navigationController?.pushViewController(viewController!, animated: true)}
        } else if let controllerName = item.controller {
            viewController = storyboard.instantiateViewController(withIdentifier: controllerName)
            showBlock = {self.navigationController?.pushViewController(viewController!, animated: true)}
        } else if let items = item.childItems {
            let tableController = storyboard.instantiateViewController(withIdentifier: "MasterTableViewController") as! MasterTableViewController
            tableController.adItems = items
            tableController.title = item.title
            showBlock = {self.navigationController?.pushViewController(tableController, animated: true)}
            viewController = tableController
        }

        if let managedController = viewController as? AppControllerManagement,
            let userData = item.userData {
            managedController.applyUserData(userData)
        }
        
        showBlock()
    }
}
